crates/holochain/src/core/workflow/app_validation_workflow.rs

```diff
@@ -95,28 +99,57 @@ async fn app_validation_workflow_inner(
 ) -> WorkflowResult<WorkComplete> {
     let env = workspace.validation_limbo.env().clone();
 
-    let ops: Vec<ValidationLimboValue> = fresh_reader!(env, |r| workspace
-        .validation_limbo
-        .drain_iter_filter(&r, |(_, vlv)| {
-            match vlv.status {
-                // We only want sys validated or awaiting app dependency ops
-                ValidationLimboStatus::SysValidated | ValidationLimboStatus::AwaitingAppDeps(_) => {
-                    Ok(true)
-                }
-                ValidationLimboStatus::Pending | ValidationLimboStatus::AwaitingSysDeps(_) => {
-                    Ok(false)
-                }
-            }
-        })?
-        // TODO: Sort into min heap like sys val
-        .collect())?;
+    // Drain the ops into a sorted binary heap
+    let sorted_ops: BinaryHeap<std::cmp::Reverse<OrderedOp<ValidationLimboValue>>> =
+        fresh_reader!(env, |r| {
+            let validation_limbo = &mut workspace.validation_limbo;
+            let element_pending = &workspace.element_pending;
+
+            let sorted_ops: Result<
+                BinaryHeap<std::cmp::Reverse<OrderedOp<ValidationLimboValue>>>,
+                WorkflowError,
+            > = validation_limbo
+                .drain_iter_filter(&r, |(_, vlv)| {
+                    match vlv.status {
+                        // We only want sys validated or awaiting app dependency ops
+                        ValidationLimboStatus::SysValidated
+                        | ValidationLimboStatus::AwaitingAppDeps(_) => Ok(true),
+                        ValidationLimboStatus::Pending
+                        | ValidationLimboStatus::AwaitingSysDeps(_) => Ok(false),
+                    }
+                })?
+                .map_err(WorkflowError::from)
+                .map(|vlv| {
+                    // Sort the ops into a min-heap
+                    let op = light_to_op(vlv.op.clone(), element_pending)?;
+
+                    let hash = DhtOpHash::with_data_sync(&op);
+                    let order = DhtOpOrder::from(&op);
+                    let v = OrderedOp {
+                        order,
+                        hash,
+                        op,
+                        value: vlv,
+                    };
+                    // We want a min-heap
+                    Ok(std::cmp::Reverse(v))
+                })
+                .iterator()
+                .collect();
+            sorted_ops
+        })?;
 
     // Validate all the ops
-    for mut vlv in ops {
+    for so in sorted_ops {
```

Comment:
I think you may need to consume the heap for iteration to happen in sort order:

[`iter`](https://doc.rust-lang.org/std/collections/struct.BinaryHeap.html#method.iter) is in arbitrary order, but [`into_iter_sorted`](https://doc.rust-lang.org/std/collections/struct.BinaryHeap.html#method.into_iter_sorted) looks good for this

```suggestion
    for so in sorted_ops.into_iter_sorted() {
```

Reply:
Dam and that method is experimental so we can't use it. We could `into_sorted_vec` which has the same `O(n * log(n))` in place cost as unstable sort on a `Vec`.
I just did a real simple experiment between binary heap, sorting a vec and btreeset because I couldn't tell from the docs.
It really doesn't make that much difference at all. Plus I doubt we will be dealing with large limbos anyway.
Looks like the priority queue is slightly faster even with `into_vec_sorted()` so I'll stick with that for now but happy to change to sorted vec because it's easier to understand.

(pq is priority queue / binary heap)
```
10 items bt took 0s, 0ms 24us 24778ns
10 items vec took 0s, 0ms 23us 23323ns
10 items pq took 0s, 0ms 26us 26438ns
100 items bt took 0s, 1ms 1151us 1151425ns
100 items vec took 0s, 1ms 1495us 1495585ns
100 items pq took 0s, 1ms 1187us 1187353ns
1000 items bt took 0s, 31ms 31830us 31830667ns
1000 items vec took 0s, 32ms 32559us 32559983ns
1000 items pq took 0s, 25ms 25702us 25702389ns
10000 items bt took 0s, 503ms 503591us 503591771ns
10000 items vec took 0s, 485ms 485417us 485417824ns
10000 items pq took 0s, 417ms 417083us 417083738ns
100000 items bt took 6s, 6891ms 6891228us 6891228336ns
100000 items vec took 6s, 6243ms 6243607us 6243607694ns
100000 items pq took 5s, 5909ms 5909932us 5909932277ns
```